File: haskell/src/Starter.hs
module Starter
  ( projectName
  , add
  , greet
  ) where

import Data.Char (isSpace)
import Data.List (dropWhileEnd)

projectName :: String
projectName = "starter"

add :: Int -> Int -> Int
add a b = a + b

greet :: String -> String
greet name =
  let who = case trim name of
        "" -> "world"
        trimmed -> trimmed
   in "Hello, " ++ who

trim :: String -> String
trim = dropWhileEnd isSpace . dropWhile isSpace

File: cpp/src/starter.cpp
#include "starter.hpp"

namespace starter {

const char *project_name() {
    return "starter";
}

int add(int a, int b) {
    return a + b;
}

std::string greet(const std::string &name) {
    const auto begin = name.find_first_not_of(" \t\n\r\f\v");
    if (begin == std::string::npos) {
        return "Hello, world";
    }
    const auto end = name.find_last_not_of(" \t\n\r\f\v");
    return "Hello, " + name.substr(begin, end - begin + 1);
}

}  // namespace starter
